### ISSUE TITLE
Possible EOL issue

### ISSUE THREAD

**Issue by ndrone** [revelc/impsort-maven-plugin#4]
Running the `check` goal it is failing on the same file over and over. My team is has a current mix of OS's Windows and Apple. For that reason we are using 'LF' line endings. 

The Imports
```
import java.util.Arrays;
import java.util.concurrent.TimeUnit;

import org.apache.commons.lang3.StringUtils;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.reflect.MethodSignature;

import com.google.common.base.Stopwatch;
```

The plugin config
```
<plugin>
    <groupId>net.revelc.code</groupId>
    <artifactId>impsort-maven-plugin</artifactId>
    <executions>
        <execution>
            <phase>validate</phase>
            <goals>
                <goal>check</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <groups>java.,javax.,org.,com.</groups>
        <staticGroups>java,*</staticGroups>
    </configuration>
</plugin>
```

any ideas on why it is failing? I took out the blank lines, and it still failed on goal execution.

Thanks,

**Comment by hazendaz** [comment]
@ndrone I'm pretty sure it touches the entire file as I've noticed end of file line endings being touched.  I have seen failures when specific character sets are used but otherwise I have not run into any specific issues in usage I've had.  I've probably got this in ~100 applicatinos at this point.  Are you able to provide more about the file in question or confirm it's all a specific character set throughout the source file?

**Comment by ctubbsii** [comment]
The most common reason I've seen for failures is that the parser fails because the file is not valid Java (does not compile). If you have the stack trace, or an example project, that could help.

**Comment by ndrone** [comment]
The source file is encoded with UTF-8 with LF line endings.  Below is the entire class in question

```
package com.example.cxp.logging;

import java.util.Arrays;
import java.util.concurrent.TimeUnit;

import org.apache.commons.lang3.StringUtils;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.reflect.MethodSignature;

import com.google.common.base.Stopwatch;


/**
 * CxpMetricsAspect is an AOP aspect. It's one method intercepts and wraps the REST API calls for
 * the purposes and logging and recording performance metrics. Included as part of Customer metrics
 * logging Created a custom annotation @LogExecutionTime Include @LogExecutionTime on top of the
 * rest methods in the controllers that needs to capture the execution time metrics
 *
 * @author
 *
 */
@Aspect
public class CxpMetricsAspect
{
    @Around("@annotation(com.example.cxp.logging.LogExecutionTime)")
    public Object timeAndLoggingExecution(ProceedingJoinPoint joinPoint) throws Throwable
    {
        // Create and start a StopWatch
        Stopwatch requestProcessTime = Stopwatch.createStarted();
        // Execute the wrapped method
        Object retval = joinPoint.proceed();
        // Stop the StopWatch
        requestProcessTime.stop();

        // grab action check to see if it's overwritten from the annotation
        MethodSignature methodSignature = (MethodSignature) joinPoint.getSignature();
        Arrays.stream(methodSignature.getMethod().getAnnotations())
            .filter(annotation -> annotation instanceof LogExecutionTime).forEach(annotation -> {

                // get the overwritten action or just use the method name
                String action = StringUtils.defaultIfBlank(((LogExecutionTime) annotation).action(),
                    joinPoint.getSignature().getName());

                // Log the metrics
                CxpMetricsUtil.trackTiming("class='{}' action='{}' elapsedTime={}",
                    joinPoint.getSignature().getDeclaringType().getName(), action,
                    requestProcessTime.elapsed(TimeUnit.MILLISECONDS));
            });
        return retval;
    }
}
```

Also, what does compile have anything to do with the check goal? I'm executing check during the validate phase of maven before the compile phase even starts. Same phase as the formatter-maven-plugin. Is this the correct phase to do this import order check?

**Comment by ctubbsii** [comment]
It doesn't technically have to compile in order to check... it just needs to be valid Java... because the Java source code parser library I use to read the file will blow up if it's not valid Java and it can't parse the file. Saying that it needs to be able to compile is just a proxy for "must be valid Java".

Executing at the validate phase is perfectly fine. It just might blow up due to invalid Java, before it gets to the compile phase of the build lifecycle, where the maven-compiler-plugin could tell you something more useful. This is a known limitation, since I'm using an external library for reading the Java source files... and didn't want to write my own Java source file parser which only read the import statements (my first pass, I actually tried to do that... but it was awful).

I can try to reproduce the problem with this provided source code later today. Just to double-check, when you say the file is encoded UTF-8, do you mean with or without the BOM. I've only tried against files without a BOM... I'm not sure if the parser handles files with one.

**Comment by ndrone** [comment]
The project compiles fine before I added the plugin to the project, so I doubt it's a compile issue. The project is a large multi-module project with multiple BOM's. Here is another project of mine that shares a similar project structure: https://github.com/ndrone/stackoverflow/tree/master/48853995/ordersystem

**Comment by ctubbsii** [comment]
Sorry, I was asking about BOM (byte-order mark... which indicates byte order for UTF encoding in files), not POM (project object model... which describes the Maven project structure, bindings, and dependencies).

**Comment by ndrone** [comment]
Sorry my mistake, how do I check the byte order mark?

**Comment by matozoid** [comment]
@ctubbsii - if JavaParser does not handle BOM's, please open an issue. If you want JavaParser to parse the imports even when the rest of the file is garbage, please open an issue. Please :-)

**Comment by hazendaz** [comment]
You can check if bom using notepad ++ and try to change encoding.  If it has bom, it will be marked.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by ndrone** [comment]
Running the following from terminal `head -c 3 CxpMetrixAspect.java | hexdump -C` I get the following result
`00000000  70 61 63                                          |pac|`

which means the file doesn't have a BOM

**Comment by ctubbsii** [comment]
The easiest thing to do to verify the problem is to run `impsort:sort` and identify what it changed. I did this and re-ran `impsort:check` and it passed. The only change was the removal of the extra blank line after all the imports.

The way the check works is by formatting the file's imports section, and reinserting it into the context of the file and checks for differences. This is a very naive way to check, but it works. When it reinserts the imports section, it does so with only a single blank line after the imports, which doesn't match the two which existed previously.

In short, I think this is caused by a normal formatting error, and does not have anything to do with line-ending issues.

(I also checked, and the parser handles BOMs just fine)

**Comment by ndrone** [comment]
yep that was it having 2 blank lines after the import block. I wonder if this still isn't a bug, though around the blank lines, what if in my eclipse formatter I have 2 blank lines after the import would this still be an issue?

**Comment by ctubbsii** [comment]
Yeah... so, I could probably patch it so it (optionally) preserves however many lines existed before sorting... or is more tolerant with the check goal... or maybe makes it configurable how many lines to place after (or between each section).

However, I'm not sure any of those options would add sufficient value to justify the dramatic increase in code complexity. I'll consider a PR to add the feature, if somebody else is willing to do it... but I don't think I'll be working on it myself, at least not any time soon.

For now, if the number of spaces is important to you, I would run the formatter-maven-plugin (or other formatting tool) after this plugin is executed.

**Comment by ndrone** [comment]
well for me I have one space specified between the package and imports and one space specified after imports... So it's not this plugin issue as far as I'm concerned. I'll look into see what it would take to give you guys a PR. BTW nice work on both plugins I'm going to be making some great use of them enforcing code stylings.
